### ISSUE TITLE
support validating responses in the UI

### ISSUE THREAD

**Issue by yoon** [NUARIG/surveyor#34]
Surveyor's Validation model now determines whether or not an answer is_valid? per the criteria set out by the author of the survey. This functionality is not yet hooked into the views, and needs to be implemented in a style similar to Question and QuestionGroup Dependencies.

**Comment by yoon** [comment]
an additional guide on inline validation in web forms:
http://www.alistapart.com/articles/inline-validation-in-web-forms/

also, a nice reminder about what to do/what not to do in forms:
http://www.alistapart.com/articles/sensibleforms/

**Comment by yoon** [comment]
http://www.smashingmagazine.com/2010/06/15/spice-up-your-website-with-jquery-goodness/

**Comment by Nataliya** [comment]
For NCS need to verify the formats of various inputs. Ex :

FORMAT MUST BE AA # # # # # # #-UR##

**Comment by hoffmanc** [comment]
So, to be clear: there is no support for validation of answers?  What about validation of a collection of answers (e.g., a_1 + a_2 + a_3 == 100)?

**Comment by yoon** [comment]
There is on the back end, but it doesn't appear in the UI yet. We don't have a syntax yet for the validation of a collection of answers. But there are dependencies that could work on a collection of answers.

**Comment by yoon** [comment]
Moving validations to the client side would have the same effects as moving dependencies to the client side - a lack of awareness of responses outside of the current section. We should build consensus on how to implement this both in the short and longer terms.

**Comment by kumbafu** [comment]
I'm hesitant to add this to the core UI. Different people will want to handle invalid answers differently. I've had two separate requests/workflows from users on how to handle invalid questions. If we do implement in the core, it should be in a way that's easy to turn off or override.

**Comment by rsutphin** [comment]
I agree with @kumbafu — Surveyor should not enforce validations unless that enforcement can be customized/disabled. It should provide a mechanism that:
- Allows defining validations for questions in the DSL
- Allows for determining if a particular response is valid/invalid
- Allows for determining the set of valid/invalid responses in a response set

It may also need to provide an extension point for displaying validation messages in the UI. I'm not quite sure how that would work; maybe one part would be including validation information in the AJAX HTTP response when a question response is submitted.

It may also be worthwhile to make the validation syntax in the DSL extensible so that applications can define their own sorts of validations.

> I've had two separate requests/workflows from users on how to handle invalid questions.

@kumbafu, it would be helpful to have the parameters for those workflows to ensure that Surveyor's validation implementation would enable (and not obstruct) them.

**Comment by kumbafu** [comment]
The two workflows I've been asked for are:
1. To be able to highlight invalid responses (but allow the form to be filled out/completed)
2. Disallow completion of forms that have invalid responses.

**Comment by yoon** [comment]
We will implement:

Client side: mandatory questions and data types (e.g. integer, date, string)
Server side: validations as defined in the DSL with rules and validation conditions

**Comment by alepore** [comment]
was this implemented?

**Comment by pfriedman** [comment]
No issue #34 has not yet been implemented.

There is a branch[1] that was started which uses Jquery.validations[2][3]
and it works for the most basic scenarios[4].

However this was just a start and there are several other issues to
address before being anywhere near Œcomplete¹.

Hopefully I, or someone, can pick it up again soon.

Kind regards, 
Paul

[1] https://github.com/NUBIC/surveyor/compare/34_support-validations
[2] http://bassistance.de/jquery-plugins/jquery-plugin-validation/
[3] http://docs.jquery.com/Plugins/Validation
[4] 
https://github.com/NUBIC/surveyor/blob/37a5e9a68e4bdfde020a7e42c0b15299e8c9
9009/lib/assets/javascripts/surveyor/surveyor_validations.js

From:  Alessandro Lepore notifications@github.com
Reply-To:  NUBIC/surveyor
<reply+i-80881-b594b9a6c5e393df7fcf71d7ca4f482c24d04d05-34622@reply.github.
com>
Date:  Saturday, May 31, 2014 at 6:34 AM
To:  NUBIC/surveyor surveyor@noreply.github.com
Cc:  Paul Friedman p-friedman@northwestern.edu
Subject:  Re: [surveyor] support validating responses in the UI (#34)

was this implemented?
‹
Reply to this email directly or
view it on GitHub 
https://github.com/NUBIC/surveyor/issues/34#issuecomment-44745907.
